### DISCUSSION TITLE
When I want to compress PDF，I got an bigger PDF

### DISCUSSION THREAD

**Discussion by Syntamin** [pymupdf/PyMuPDF#3645] [Q&A]
### Description of the bug

I want to compress a PDF. I extracted images from the PDF file and then used pngquant to compress them. Their size was reduced by more than 70%. However, when I used the replace_image function to replace the images, the size of the new PDF became bigger. I want to know why this happens, and if I used the save function incorrectly.

### How to reproduce the bug

    def compress_pdf(input_path, output_path=""):
    doc = pymupdf.open(input_path)

    doc_name_with_extension = os.path.basename(input_path)
    doc_name = os.path.splitext(doc_name_with_extension)[0]

    for page_index in range(10):
        page = doc[page_index]
        image_list = page.get_images()

        for image_index, img in enumerate(image_list, start=1):
            xref = img[0]
            pix = pymupdf.Pixmap(doc, xref)

            if pix.n - pix.alpha > 3:
                pix = pymupdf.Pixmap(pymupdf.csRGB, pix)

            origin_png_path = "./origin/%s_page_%s-image_%s.png" % (
                doc_name,
                page_index,
                image_index,
            )

            pix.save(origin_png_path)  # 存储提取出的图片
            pix = None

            pngquant.compress_png(origin_png_path)

            compressed_png_path = "./origin/%s_page_%s-image_%s-fs8.png" % (
                doc_name,
                page_index,
                image_index,
            )

            # print(os.path.getsize(compressed_png_path), compressed_png_path)

            # 1. replace to file
            # page.replace_image(xref, filename=compressed_png_path)

            with open(compressed_png_path, "rb") as compressed_png:
                compressed_png_bytes = compressed_png.read()
                print(len(compressed_png_bytes), "111")
                page.replace_image(xref, stream=compressed_png_bytes)

    doc.save(output_path, garbage=3, clean=True)
    doc.close()

### PyMuPDF version

1.24.6

### Operating system

MacOS

### Python version

3.9

**Comment by JorjMcKie** [comment]
If you replace images, there is always the possibility that the new image is larger or leads to a larger result. Especially when inserting PNGs, MuPDF uses its own compression, which is not necessarily the one with the largest compression results.
You also did not use the maximum available compressions option of the `save` method. We recommend to use `ez_save()`.
Finally, "clean" can lead to file increases and is generally counter-productive in this context.

**Reply by Syntamin** [reply]
I extracted an original image from a PDF file, which was 399KB. After compressing it with pngquant, it reduced to 112KB. However, after using `replace_image` to replace it in the PDF, the file size increased by around 100KB. Even when using `ez_save`, I got the same result. Normally, I expected the PDF size to decrease by about 200KB. Could there be an issue with the parameters I used for saving? What could potentially cause this outcome?

**Reply by JorjMcKie** [reply]
Does the image have a transparency channel (alpha)?
If not you may want to convert it to JPG which often leads to smaller sizes.

In addition, try `doc.ez_save(filename, compression_effort=100)`. This **_may_** lead to smaller sizes because MuPDF uses the most it can to compress compressible objects.

**Reply by Syntamin** [reply]
Thanks for your suggestion. I previously used JPEG format images and lowered the image quality, which indeed effectively reduced the file size. The reason I chose PNG over JPEG is that I wanted to perform lossless compression, maximizing size reduction without sacrificing image clarity.

Additionally, I used doc.ez_save(filename, compression_effort=100). Specifically, the original file size was 20MB, and saving the original file with ez_save reduces it to 19MB. However, if I replace the compressed images and then save it with ez_save, the file becomes 19.5MB. I think there might be some additional file information added in between. My understanding is that replacing the original embedded images with smaller images and deleting the original image objects should result in a smaller file size, but that’s not the case.

Do you know the reason behind this? Or do you have any suggestions for reducing the file size when replacing the original images with smaller ones of the same format? My ultimate goal is to reduce the size of images, text, fonts, and metadata to maximize the reduction of the PDF file size while maintaining clarity.

**Reply by JorjMcKie** [reply]
I tried to explain that further up: MuPDF uses its own mechanism to store PNG - no matter what tool you used to create a super-compressed PNG version.

**Reply by JorjMcKie** [reply]
So it may end up with a larger PNG inside the PDF than it was before insertion as a separate image.
Try to convert the PNG to a JPG - this may have better results.

**Reply by Syntamin** [reply]
Thank you very much for your explanation, I understand now.
